v2/robotmk/src/schtasks.rs

```diff
@@ -0,0 +1,265 @@
+            .checked_sub_signed(ChronoDuration::minutes(1))
+            .context("Failed to compute value for start time option in task creation command")?
+            .format("%H:%M")
+            .to_string(),
+        "/f",
+    ])?;
+    Ok(())
+}
+
+fn build_task_script(command_spec: &CommandSpec, paths: &Paths) -> String {
+    [
+        String::from("@echo off"),
+        format!(
+            "powershell.exe (Get-WmiObject Win32_Process -Filter ProcessId=$PID).ParentProcessId > {}",
+            paths.pid
+        ),
+        format!("{command_spec} > {} 2> {}", paths.stdout, paths.stderr),
+        format!("echo %errorlevel% > {}", paths.exit_code),
+    ]
+    .join("\n")
+}
+
+fn run_schtasks<T>(arguments: impl IntoIterator<Item = T>) -> Result<String>
+where
+    T: AsRef<str>,
+{
+    let mut command = Command::new("schtasks.exe");
+    command.args(arguments.into_iter().map(|a| a.as_ref().to_string()));
+    let output = command
+        .output()
+        .context(format!("Failed to run schtasks: {command:?}"))?;
+    if !output.status.success() {
+        bail!(format!(
+            "schtasks exited non-successfully. Command:\n{command:?}\n\nStdout:\n{}\n\nStderr:\n{}",
+            String::from_utf8_lossy(&output.stdout),
+            String::from_utf8_lossy(&output.stderr)
+        ))
+    }
+    String::from_utf8(output.stdout.clone()).context(format!(
+        "Failed to decode stdout of schtasks. Command:\n{command:?}\n\nLossy stdout:\n{}",
+        String::from_utf8_lossy(&output.stdout)
+    ))
+}
+
+async fn wait_for_task_exit(
+    task_name: &str,
+    termination_flag: &TerminationFlag,
+    path_pid: &Utf8Path,
+) -> Result<Option<RunOutcome>> {
+    debug!("Waiting for task {} to complete", task_name);
+    while query_if_task_is_running(task_name)
+        .context(format!("Failed to query if task {task_name} is running"))?
+    {
+        if termination_flag.should_terminate() {
+            kill_and_delete_task(task_name, path_pid);
+            return Ok(Some(RunOutcome::Terminated));
+        }
+        sleep(Duration::from_millis(250)).await
+    }
+    Ok(None)
+}
+
+fn query_if_task_is_running(task_name: &str) -> Result<bool> {
+    let schtasks_stdout = run_schtasks(["/query", "/tn", task_name, "/fo", "CSV", "/nh"])?;
+    Ok(schtasks_stdout.contains("Running"))
+}
+
+fn kill_and_delete_task(task_name: &str, path_pid: &Utf8Path) {
+    error!("Killing and deleting task {task_name}");
+    kill_task(path_pid);
+    delete_task(task_name);
+}
+
+fn kill_task(path_pid: &Utf8Path) {
+    let raw_pid = match read_from_file_before_first_whitespace(path_pid) {
+        Ok(raw_pid) => raw_pid,
+        Err(err) => {
+            log_and_return_error(err);
+            return;
+        }
+    };
+    let pid = match Pid::from_str(&raw_pid).context(format!("Failed to parse {} as PID", raw_pid)) {
+        Ok(pid) => pid,
+        Err(err) => {
+            log_and_return_error(err);
+            return;
```

Comment:
This would be worth a panic.

Reply:
Yes, we could indeed panic here. Let's leave it as is for now, we can always change it if necessary. At least, the error isn't completely lost, since we log it.